### PR TITLE
rustdoc: Sort implementors

### LINKED ISSUE #53812: doc ordering seems random
https://doc.rust-lang.org/std/iter/trait.IntoIterator.html#implementors

impl<'a, T> IntoIterator for &'a mut [T; 12]
impl<'a, T> IntoIterator for &'a [T; 25]
impl<'a, T> IntoIterator for &'a mut [T; 21]
impl<'a, T> IntoIterator for &'a [T; 2]
...

this ordering doesn't make any sense and makes the docs harder to navigate.

### PR DESCRIPTION
**Pull request by kzys**
Fixes #53812

### CODE REVIEW

**Comment by kzys** [comment]
Sorry, it shouldn't have the submodules. The latest commit doesn't have them.

**Comment by GuillaumeGomez** [comment]
Thanks!

The lack of test is problematic though, but you can't really add one since the UI tests are still pending... Should we wait the merge of the UI tests PR @QuietMisdreavus?

**Comment by TimNN** [comment]
Ping from triage @GuillaumeGomez: It looks like we won't be able to run the UI tests on CI soon. Do you still want to mark this PR as blocked on it? Otherwise, this PR requires your review.

**Comment by kzys** [comment]
How about adding tests under `src/test/rustdoc/`? Seems like I can validate something like "the first element under this element must be X".

**Comment by QuietMisdreavus** [comment]
Testing something like "the first element in this list must be X" should fit into a `src/test/rustdoc` test, since the XPath implementation we use should be able to support that:

https://github.com/rust-lang/rust/blob/994cdd918589535d705177545bf503cd0c3c5148/src/etc/htmldocck.py#L67-L97

I'd be okay with a test that did that.

**Comment by kzys** [comment]
Hmm, is there a way to have the test file as std?

```rust
pub trait MyIterator {
}

macro_rules! array_impls {
    ($($N:expr)+) => {
        $(
            impl<'a, T> MyIterator for &'a mut [T; $N] {
            }
        )+
    }
}

array_impls! { 0 1 2 }
```

The below build as its own crate, which makes all MyIterator implementations as [Implementations on Foreign Types](https://github.com/rust-lang/rust/pull/43849). And the section somehow doesn't have this problem. In other words, I can't repro the issue in this test.

**Comment by QuietMisdreavus** [comment]
Try using a wrapper type instead of plain arrays:

```rust
pub trait MyIterator {
}

pub struct MyStruct<T>(T);

macro_rules! array_impls {
    ($($N:expr)+) => {
        $(
            impl<'a, T> MyIterator for MyStruct<&'a mut [T; $N]> {
            }
        )+
    }
}

array_impls! { 0 1 2 }
```

That should make rustdoc treat the implementation as on a local type.

File: src/test/rustdoc/issue-53812.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+macro_rules! array_impls {
+    ($($N:expr)+) => {
+        $(
+            impl<'a, T> MyIterator for &'a MyStruct<[T; $N]> {
+            }
+        )+
+    }
+}
+
+// @has issue_53812/trait.MyIterator.html '//*[@id="implementors-list"]//h3[1]' 'MyStruct<[T; 0]>'
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/issue-53812.rs]: Please check `h3[2]`, `h3[3]` and `h3[4]` too. Like that we're **sure** it's not just luck.

**Pull request merged**